cmd/promtool/debug.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/google/pprof/profile"
+)
+
+type debugWriterConfig struct {
+	server         string
+	tarballName    string
+	pathToFileName map[string]string
+	postProcess    func(b []byte) (*bytes.Buffer, error)
+}
+
+type debugWriter struct {
+	archiver
+	httpClient
+	requestToFile map[*http.Request]string
+	postProcess   func(b []byte) (*bytes.Buffer, error)
+}
+
+func newDebugWriter(cfg debugWriterConfig) *debugWriter {
+	client, err := newHTTPClient(httpClientConfig{serverURL: cfg.server})
+	if err != nil {
+		panic(err)
```

Comment:
Panics should only be used to catch programming errors, otherwise we want to output a clean error message instead and return with the right exit code (`1`). In this case, `newHTTPClient` can fail when the user-provided server URL cannot be parsed.

Reply:
I see! Thanks for reminding!

Reply:
Should [here](https://github.com/chyeh/prometheus/blob/PROM-1551/cmd/promtool/main.go#L356) be the same?

Reply:
A panic is fine there because the only way to get there is a programming error (not a user error, not a transient network or IO error or such).

Reply:
I see!! Thanks for the explantation!!